packages/discovery-provider/plugins/pedalboard/apps/backfill-audio-analyses/src/backfill_discovery.ts

```diff
@@ -0,0 +1,223 @@
+      })
+      if (response.status == 200) {
+        console.log(
+          `Successfully retrieved audio analysis for track ID ${
+            track.track_id
+          }, track CID ${trackCid}${
+            audioUploadId ? `, upload ID: ${audioUploadId}` : ''
+          } via ${contentNode}`
+        )
+        const resultsKey = isLegacyTrack ? 'results' : 'audio_analysis_results'
+        const errorCountKey = isLegacyTrack
+          ? 'error_count'
+          : 'audio_analysis_error_count'
+        const results = response.data[resultsKey]
+        const errorCount = response.data[errorCountKey]
+        if (!results) {
+          break
+        }
+
+        let musicalKey = null
+        let bpm = null
+        if (results?.key) {
+          musicalKey = results?.key
+        } else if (results?.Key) {
+          musicalKey = results?.Key
+        }
+        if (results?.bpm) {
+          bpm = results?.bpm
+        } else if (results?.BPM) {
+          bpm = results?.BPM
+        }
+
+        result = {
+          track_id: track.track_id,
+          musical_key: musicalKey,
+          bpm,
+          error_count: errorCount
+        }
+        break
+      } else {
+        console.log(
+          formatErrorLog(
+            `Received ${response.status} response`,
+            track,
+            contentNode,
+            i + 1
+          )
+        )
+        continue
+      }
+    } catch (error: any) {
+      console.log(formatErrorLog(error.message, track, contentNode, i + 1))
+      continue
+    }
+  }
+
+  release() // release the semaphore permit
+  return result
+}
+
+async function fetchTracks(
+  offset: number,
+  limit: number,
+  db: Knex
+): Promise<Track[]> {
+  return await db<Track>('tracks')
+    .where(function () {
+      this.whereNull('musical_key').orWhereNull('bpm')
+    })
+    .andWhere('audio_analysis_error_count', '<', 3)
+    .andWhere('track_cid', 'is not', null)
+    .orderBy('track_id', 'asc')
+    .offset(offset)
+    .limit(limit)
+}
+
+// trigger audio analyses for batchSize tracks at a time
+async function processBatches(db: any, batchSize: number): Promise<void> {
+  let offset
+  while (true) {
+    console.time('Batch processing time')
+    const contentNodes = await getCachedHealthyContentNodes()
+    if (contentNodes.length == 0) {
+      console.timeEnd('Batch processing time')
+      console.error(`No healthy content nodes found. Please investigate`)
+      return
+    }
+    offset = await readDbOffset(DB_OFFSET_KEY)
+    if (offset == null) {
+      offset = 0
+    }
+    const tracks = await fetchTracks(offset, batchSize, db)
+    if (tracks.length === 0) {
+      console.timeEnd('Batch processing time')
+      break
+    }
+    const analyzePromises = tracks.map((track) =>
+      getAudioAnalysis(
+        contentNodes.map((node) => node.endpoint),
+        track
+      )
+    )
+    const updates = (await Promise.all(analyzePromises)).filter(
+      (u) => u !== null
```

Comment:
so getAudioAnalysis may return undefined in the catch statement right? so would we want to just check truthyness?

Reply:
it actually should catch and try the next node until there are no more nodes, so it should always return null if all nodes fail, but i changed it to `!= null` just in case

Reply:
ah nice, thanks